File: src/main/java/me/oriharel/machinery/utilities/RandomCollection.java
package me.oriharel.machinery.utilities;

import java.util.NavigableMap;
import java.util.Random;
import java.util.TreeMap;

/**
 * Get random object from collection with weight
 *
 * @param <E> type of collection
 */
public class RandomCollection<E> {
    private final NavigableMap<Double, E> map = new TreeMap<Double, E>();
    private final Random random;
    private double total = 0;

    public RandomCollection() {
        this(new Random());
    }

    public RandomCollection(Random random) {
        this.random = random;
    }

    public RandomCollection<E> add(double weight, E result) {
        if (weight <= 0) return this;
        total += weight;
        map.put(total, result);
        return this;
    }

    public boolean isPresent(double weight) {
        return map.containsKey(weight);
    }

    public E get(double weight) {
        return map.get(weight);
    }

    public void clear() {
        total = 0;
        map.clear();
    }

    public E next() {
        double value = random.nextDouble() * total;
        return map.higherEntry(value).getValue();
    }
}

File: src/main/java/me/oriharel/machinery/utilities/RandomCollection.kt
package me.oriharel.machinery.utilities

import java.util.*

/**
 * Get random object from collection with weight
 *
 * @param <E> type of collection
</E> */
class RandomCollection<E> @JvmOverloads constructor(private val random: Random = Random()) {
    private val map: NavigableMap<Double, E> = TreeMap()
    private var total = 0.0
    fun add(weight: Double, result: E): RandomCollection<E> {
        if (weight <= 0) return this
        total += weight
        map[total] = result
        return this
    }

    fun isPresent(weight: Double): Boolean {
        return map.containsKey(weight)
    }

    operator fun get(weight: Double): E? {
        return map[weight]
    }

    fun clear() {
        total = 0.0
        map.clear()
    }

    operator fun next(): E {
        val value = random.nextDouble() * total
        return map.higherEntry(value).value
    }

}